### PR TITLE
Use cortex as data source

### PR DESCRIPTION
**Pull request by marians**
Towards https://github.com/giantswarm/giantswarm/issues/16616

By using Cortex (Prometheus) as a single data source for the existence of test clusters, we no longer need a REST API account for resource-police with admin permissions in all test installations (which is a deprecated concept).

**CAVEAT**: From cortex we currently get a lot less detail on clusters than we did via the REST API. E. g. no cluster name, creator, creation date. I hope that's something we can improve in the future, either by adding to prometheus labels, or through other means (e. g. going through the Management API).

The implementation relies on the `aggregation:giantswarm:cluster_release_version` series.

## Preview

image

## TODO

- Test in production

### CODE REVIEW

File: pkg/cortex/cortex.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+	value, warnings, err := v1api.Query(context.Background(), query, t)
+	if err != nil {
+		return clusters, microerror.Mask(err)
+	}
+	if len(warnings) > 0 {
+		fmt.Printf("Warnings while querying: %#v\n", warnings)
+	}
+
+	if value == nil {
+		return clusters, microerror.Mask(fmt.Errorf("query %s returned nil value for time %s", query, t))
```
Comment (rossf7) [inline pkg/cortex/cortex.go:77]:
```suggestion
		return clusters, microerror.Maskf(executionFailedError, "query %s returned nil value for time %s", query, t)
```

Could we use our regular error handling here? Just needs an executionFailedError to be definted.

File: pkg/cortex/cortex.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+)
+
+const (
+	query = `aggregation:giantswarm:cluster_release_version{pipeline="testing"}`
+)
+
+type Config struct {
+	// Base URL of the prometheus API server.
+	URL string
+	// User name.
```
Comment (rossf7) [inline pkg/cortex/cortex.go:26]:
```suggestion
	// UserName of the prometheus API server.
```

File: pkg/cortex/cortex.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+	"github.com/prometheus/common/config"
+	"github.com/prometheus/common/model"
+)
+
+const (
+	query = `aggregation:giantswarm:cluster_release_version{pipeline="testing"}`
+)
+
+type Config struct {
+	// Base URL of the prometheus API server.
```
Comment (rossf7) [inline pkg/cortex/cortex.go:24]:
```suggestion
	// URL of the prometheus API server.
```

These need to start with the variable name to be recognised as godoc.

File: pkg/intersect/intersect.go [hunk 4]
```diff
@@ -0,0 +1,20 @@
+// Package intersect is a version of https://github.com/juliangruber/go-intersect/blob/master/intersect.go
```
Comment (rossf7) [inline pkg/intersect/intersect.go:1]: Nice thanks for adding the package godoc and the link. 👍

**Review by rossf7 — COMMENTED** [review]
@marians Thanks this is great! Now we have cortex much better to fetch from there. As otherwise we're playing whackamole having working API creds for each test installation.

> From cortex we currently get a lot less detail on clusters than we did via the REST API. E. g. no cluster name, creator, creation date. I hope that's something we can improve in the future, either by adding to prometheus labels, or through other means (e. g. going through the Management API).

Setting the cluster ID like Nikola is doing is pretty easy with `kubectl gs`. I think we could ask people to do that if they want to keep their clusters.

But I agree having a prometheus label with the cluster name would help. I think that can be done as a later step.

**Comment by marians** [comment]
Thanks @rossf7 ! Just pushed changes according to your suggestions.

**Comment by marians** [comment]
I just recalled that we have one more detail that we could display in the report alredy: the release version the cluster is using. Sometimes that includes hints on the purpose of the cluster or the creator. E. g. 

image

**Review by rossf7 — APPROVED** [review]
LGTM

**Comment by marians** [comment]
I changed the app on gorilla/rfjh2 to use the dev version. Let's wait for the cronjob and see.

**Comment by marians** [comment]
FYI: I just removed the "Please check our policy ..." hint from the Slack message, as resource-police in this version does not take labels into account.

**Pull request merged**